### PR TITLE
Incline rocm 5.6.0 dev

### PR DESCRIPTION
**Pull request by cameronrutherford**

### CODE REVIEW

File: .gitlab/pnnl/base.gitlab-ci.yml [hunk 1]
```diff
@@ -269,4 +269,4 @@ stages:
   variables:
     WORKDIR_SUFFIX: "x86_64-clang-hip-build"
     MY_CLUSTER: "incline"
-    SLURM_ARGS: " --exclusive --ntasks=3 "
+    SLURM_ARGS: " -N 1 --ntasks=3 "
```
Comment (cameronrutherford) [inline .gitlab/pnnl/base.gitlab-ci.yml:272]: These arguments were sufficient when running an interactive shell to get access to GPUs, so not sure why this isn't working for CI

Comment (ryandanehy) [reply .gitlab/pnnl/base.gitlab-ci.yml:272]: That's weird. Not sure why you would need this but maybe try the --gpus flag.

File: buildsystem/spack/incline/install.sh [hunk 2]
```diff
@@ -12,5 +12,8 @@
 export MY_CLUSTER=incline
 . buildsystem/load-spack.sh &&
 spack develop --no-clone --path=$(pwd) resolve@develop &&
+spack concretize -f &&
+spack install -j 64 llvm-amdgpu &&
+spack load llvm-amdgpu &&
```
Comment (cameronrutherford) [inline buildsystem/spack/incline/install.sh:17]: @ryandanehy when building llvm-amdgpu from source with spack, this is how you get the compiler working in combination with changes in spack.yaml

File: buildsystem/spack/incline/spack.yaml [hunk 3]
```diff
@@ -40,3 +66,15 @@ spack:
       operating_system: centos7
       target: x86_64
       modules: [gcc/8.4.0]
+  - compiler:
+      spec: clang@16.0.0-rocm5.6.0
+      paths:
+        cc: amdclang
+        cxx: amdclang++
+        f77: /share/apps/gcc/8.4.0/bin/gfortran
```
Comment (cameronrutherford) [inline buildsystem/spack/incline/spack.yaml:74]: Here we pray amdclang is on the path @ryandanehy

File: resolve/LinSolverDirectRocSolverRf.hpp [hunk 4]
```diff
@@ -8,7 +8,6 @@
 #include <rocblas/rocblas.h>
 #include <rocsolver/rocsolver.h>
 #include <hip/hip_runtime.h>
-#include <roctracer/roctx.h>
```
Comment (cameronrutherford) [inline resolve/LinSolverDirectRocSolverRf.hpp:11]: Spack installation of roctx sucks. There is no exported CMake configuration to consume. I was able to install the libs with packages `roctracer-dev` and `roctracer-dev-api`, but it wasn't immediately obvious of a good way to support this without an obvious hack

File: cmake/ReSolveConfig.cmake.in [hunk 5]
```diff
@@ -15,14 +19,19 @@ if(@RESOLVE_USE_CUDA@)
...
   find_package(hip REQUIRED)
-  find_package(hipblas REQUIRED)
+  find_package(rocblas REQUIRED)
+  find_package(rocsparse REQUIRED)
+  find_package(rocsolver REQUIRED)
   # This is just an agly hack to make HIP build work
   get_target_property(hip_includes hip::device INTERFACE_INCLUDE_DIRECTORIES)
-  message(STATUS "HIP include directories: ${hip_includes}")
-  include_directories(${hip_includes})
+  target_include_directories(ReSolve::resolve_hip INTERFACE $<BUILD_INTERFACE:${hip_includes}>)
```
[the remark this answers is not in the document]
Comment (cameronrutherford) [reply cmake/ReSolveConfig.cmake.in:34]: Although this is still a hack...

File: buildsystem/spack/incline/spack.yaml [hunk 6]
```diff
@@ -1,10 +1,35 @@
...
+      modules: [gcc/8.4.0]
+  - compiler:
+      spec: clang@16.0.0-rocm5.6.0
+      paths:
+        cc: amdclang
+        cxx: amdclang++
+        f77: /share/apps/gcc/8.4.0/bin/gfortran
+        fc: /share/apps/gcc/8.4.0/bin/gfortran
+      flags:
+        cxxflags: --gcc-toolchain=/share/apps/gcc/8.4.0
```
Comment (cameronrutherford) [inline buildsystem/spack/incline/spack.yaml:29]: FYI spack refuses to function if there is a training slash here (` --gcc-toolchain=/share/apps/gcc/8.4.0/`)

File: cmake/ReSolveFindHipLibraries.cmake [hunk 7]
```diff
@@ -6,13 +6,20 @@ add_library(resolve_hip INTERFACE)
...
 
-target_link_libraries(resolve_hip INTERFACE
+target_link_libraries(resolve_hip INTERFACE 
   hip::host 
   hip::device
   roc::rocblas
   roc::rocsparse
-  rocsolver
+  roc::rocsolver
 )
```
Comment (pelesh) [inline cmake/ReSolveFindHipLibraries.cmake:17]: I had to remove `roc::` namespace in order to build on Frontier. I am not sure what is going on with HIP find functions, but it seems CMake is still a bit buggy.

```suggestion
  rocsolver
)
```

File: resolve/cpu/CMakeLists.txt [hunk 8]
```diff
@@ -17,12 +17,7 @@ set(ReSolve_CPU_HEADER_INSTALL
 
 # First create dummy backend
 add_library(resolve_backend_cpu SHARED ${ReSolve_CPU_SRC})
-target_link_libraries(resolve_backend_cpu PRIVATE resolve_logger)
-
-target_include_directories(resolve_backend_cpu INTERFACE
-    $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>
-    $<INSTALL_INTERFACE:include>
-)
+target_link_libraries(resolve_backend_cpu PUBLIC resolve_logger)
```
Comment (pelesh) [inline resolve/cpu/CMakeLists.txt:20]: Why remove include dirs here?

File: resolve/cuda/CMakeLists.txt [hunk 9]
```diff
@@ -27,10 +27,7 @@ set_source_files_properties(${ReSolve_CUDA_SRC} PROPERTIES LANGUAGE CUDA)
 add_library(resolve_backend_cuda SHARED ${ReSolve_CUDA_SRC})
 target_link_libraries(resolve_backend_cuda PRIVATE resolve_logger)
 target_link_libraries(resolve_backend_cuda PUBLIC resolve_cuda)
-target_include_directories(resolve_backend_cuda INTERFACE
-    $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>
-    $<INSTALL_INTERFACE:include>
-)
```
Comment (pelesh) [inline resolve/cuda/CMakeLists.txt:30]: Why remove include dirs here (too)?

File: resolve/cpu/CMakeLists.txt (hunk 8 above)
Comment (pelesh) [inline resolve/cpu/CMakeLists.txt:20]: Could we make logger dependency private?
```suggestion
target_link_libraries(resolve_backend_cpu PRIVATE resolve_logger)
```

File: resolve/cuda/CMakeLists.txt (hunk 9 above)
Comment (cameronrutherford) [reply resolve/cuda/CMakeLists.txt:30]: I don't know why these were added in the first place... Include directory for `./resolve` is propogated through `resolve_logger`. Perhaps some re-naming would be apt here/consolidating with workspaces?

**Pull request merged**